var/spack/repos/builtin/packages/mpich/package.py

```diff
@@ -41,11 +42,28 @@ class Mpich(AutotoolsPackage):
     version('3.1.1', '40dc408b1e03cc36d80209baaa2d32b7')
     version('3.1',   '5643dd176499bfb7d25079aaff25f2ec')
     version('3.0.4', '9c5d5d4fe1e17dd12153f40bc5b6dbc0')
+    version('develop', git='git://github.com/pmodels/mpich')
 
     variant('hydra', default=True,  description='Build the hydra process manager')
     variant('pmi',   default=True,  description='Build with PMI support')
     variant('romio', default=True,  description='Enable ROMIO MPI I/O implementation')
     variant('verbs', default=False, description='Build support for OpenFabrics verbs.')
+    variant(
+        'device',
+        default='ch3',
+        description='Abstract Device Interface (ADI) implementation',
+        values=('ch3', 'ch4'),
```

Comment:
I'm looking at this more and I have two thoughts:

* It appears that a user wouldn't set `device=ch3,ch4` (at least in terms of how the `with_or_without` functions you added work, in which case `multi=True` should be removed for this option.
* Most netmods imply the device (everything except `ofi` it appears). If a user sets `netmod=ucx` without specifying the device, then at the moment Spack would report a conflict. IMO this could be avoided if you set the default to `None` and then checked the values in the `with_or_without` methods like `if self.spec.variants['device'].value == None`. Does that seem reasonable or do you think a user typically wants to control the ADI and network module?

Sorry about making an additional request after the LGTM.

Reply:
Correct, only one device implementation is allowed at a time. I will set `multi=False`.

At the moment, we would prefer users know they are opting into the `ch4` device for 2 reasons.
* potential bugs: `ch4` is still in "experimental" state
* marketing :smile:

Reply:
Hmm, maybe I should also change this to actually support multiple netmods (which is allowed).

Reply:
> Hmm, maybe I should also change this to actually support multiple netmods (which is allowed).

If it doesn't by default then IMO you don't have to. Although IMO it would be worthwhile to mention it in a comment that the package.py doesn't support multiple netmods. With the current implementation using the `with_or_without` methods, a spec like `mpich netmod=ofi,mxm` would generate the options `--with-device=ch3:nemesis:ofi` and `--with-device=ch3:nemesis:mxm`. Would only one of these have an effect? If you want to support combinations, then IMO renaming the functions to something other than `with_or_without` would be the way to go (since functions of that form return their results independently).

> At the moment, we would prefer users know they are opting into the ch4 device for 2 reasons.

The experimental part makes sense. Would it be worth mentioning that in the description for this variant?

Reply:
Sure, I will add a note about it. I will also set `multi=False` for netmod configurations for now.